Project implementation

pmiakish opened the pull request.

pmiakish committed:
Removes redundant properties

pmiakish committed:
Add localization

pmiakish committed:
Add JSP status messages localization

pmiakish committed:
Add removal invalid locale-object from session

pmiakish committed:
Fix alerts localization in JSP

pmiakish committed:
Add servlets for displaying payments history

Localize login-page

pmiakish committed:
Add README.md, replaces OrderStrategy to dao from services

pmiakish committed:
Fix typo

pmiakish committed:
Fix typos in README.md

pmiakish committed:
Fix person removal

zennick commented on pom.xml:
```diff
@@ -79,21 +77,13 @@
-    <dependency>
-      <groupId>org.apache.maven.plugins</groupId>
-      <artifactId>maven-compiler-plugin</artifactId>
-      <version>3.8.1</version>
```
it's better to have all versions as a properties, not just log4j

zennick commented on src/main/resources/log4j2.xml:
```diff
@@ -25,6 +25,9 @@
         <Logger name="com.epam.upskillproject.controller.filters.UserFilter" level="TRACE" additivity="true">
             <AppenderRef ref="RollingFile"/>
         </Logger>
+        <Logger name="com.epam.upskillproject.controller.LocaleDispatcher" level="TRACE" additivity="true">
```
it's not used to list each class in logging config, it can remain as is, but generally nobody doing that.

zennick commented on src/main/resources/sql.properties:
```diff
@@ -61,11 +61,11 @@ query.income.increase=UPDATE INCOME SET BALANCE=BALANCE+?
 query.income.decrease=UPDATE INCOME SET BALANCE=BALANCE-?
 query.payment.getSingleById=SELECT * FROM TRANSACTION WHERE ID=?
 query.payment.getAll=SELECT * FROM TRANSACTION ORDER BY %s
-query.payment.getByPayer=SELECT * FROM TRANSACTION WHERE PAYER=? ORDER BY ID DESC
```
No index exists in database schema for such query. It works quick just because of small database volume. Need to have indexes for each query and foreign key.

pmiakish committed:
Fix custom connection pool

pmiakish committed:
Add fixes

Change HSQLDB to MySQL
Add indexes to database tables
Replaces validation methods to ParamsValidator
Fix issue with closing statements in QueryExecutor
Add length-parameters to input tags in JSPs
Fix issue with EJBException wrapping

pmiakish committed:
Add fixes

Refactor controller (implements Command-design)
Adapt syntax of SQL-queries to MySQL dialect
Fix params in JSPs
Optimize dependency versioning in POM.xml
Optimize exception handling

zennick commented on src/main/java/com/epam/upskillproject/controller/command/ActionFactory.java:
```diff
@@ -0,0 +1,238 @@
+        if (req != null && resp != null) {
+            try {
+                Optional<TargetType> target = paramReader.readTargetTypeParamOrAttr(req, TARGET_PARAM);
+                if (target.isPresent()) {
```
if not presented - return null. Should return an exception?

zennick commented on src/main/java/com/epam/upskillproject/controller/command/impl/admin/AccountActivateCommand.java:
```diff
@@ -0,0 +1,70 @@
+    private static final String OPERATION_NAME_ATTR = "opName";
+    private static final String OPERATION_STATUS_ATTR = "opStat";
+
+    private static final PermissionType[] permissions = {PermissionType.SUPERADMIN, PermissionType.ADMIN};
```
There isa terminology issue - it's a required role, not a required permission.

zennick commented on src/main/java/com/epam/upskillproject/connect/customds/CustomPooledDataSource.java:
```diff
@@ -147,23 +151,27 @@ public void setInactivityTimeLimitMillis(int inactivityTimeLimitMillis) throws I
+        try {
+            PoolConnection connection = connectionPool.poll(requestTimeoutValue, requestTimeoutUnit);
+            logger.log(Level.TRACE, "Connection obtained from connection pool");
+            if (connection == null) {
```
"poll method returns the head of this queue, or null if the specified waiting time elapses before an element is available" - need to raise an exception is can't get a connection from pool per time out - opening new can lead to the problem.

zennick commented on src/main/java/com/epam/upskillproject/connect/customds/CustomPooledDataSource.java:
```diff
@@ -172,35 +180,70 @@ public Connection getConnection(String username, String password) throws Unsuppo
     }
 
+    public void shutdown() throws SQLException {
+        for (Connection connection : activeConnections) {
```
can be a problem to close connection in the middle of operation. Why not to wait for a little and give a chance to return connection back to the pool?

zennick commented on src/main/java/com/epam/upskillproject/connect/customds/CustomPooledDataSource.java:
```diff
@@ -172,35 +180,70 @@ public Connection getConnection(String username, String password) throws Unsuppo
         try {
-            closeOldConnections();
-            if (connectionPool.size() < maxConnectionsNumber) {
-                poolConnection.setAutoCommit(true);
```
before returning to the pool - makes sense to rollback active transaction is any on connection

zennick commented on src/main/java/com/epam/upskillproject/controller/command/impl/admin/PaymentListCommand.java:
```diff
@@ -0,0 +1,72 @@
+
+    private static final Logger logger = LogManager.getLogger(PaymentListCommand.class.getName());
+
+    private static final String DATE_TIME_PATTERN = "yyyy-MM-dd HH:mm:ss";
```
Why it's not a global constant?

pmiakish committed:
Add equals(), hashCode() to Page

pmiakish committed:
Add fixes

Rename 'Permission' to 'Role' in all the classes and DB
Rename packages (in case of plural)
Add CommandNotFoundException throwing in ActionFactory
Fix connection management in connection pool
Add datetime formatter system property

pmiakish committed:
Change if-statement in ConnectionPool

pmiakish committed:
Add fixes

Fix shutdown of connection pool and connection release
Remove redundant line breaks in JSPs
